co-noir/ultrahonk/src/decider/types.rs

```diff
@@ -0,0 +1,144 @@
+use super::univariate::Univariate;
+use crate::{
+    types::{AllEntities, Polynomials},
+    NUM_ALPHAS,
+};
+use ark_ec::pairing::Pairing;
+use ark_ff::PrimeField;
+use itertools::izip;
+use std::{iter, vec};
+
+pub(crate) struct ProverMemory<P: Pairing> {
+    pub(crate) polys: AllEntities<Vec<P::ScalarField>>,
+    pub(crate) relation_parameters: RelationParameters<P::ScalarField>,
+}
+
+pub(crate) const MAX_PARTIAL_RELATION_LENGTH: usize = 7;
+pub(crate) type ProverUnivariates<F> = AllEntities<Univariate<F, MAX_PARTIAL_RELATION_LENGTH>>;
+pub(crate) type PartiallyEvaluatePolys<F> = AllEntities<Vec<F>>;
+pub(crate) type ClaimedEvaluations<F> = AllEntities<F>;
+
+pub(crate) struct RelationParameters<F: PrimeField> {
+    pub(crate) eta_1: F,
+    pub(crate) eta_2: F,
+    pub(crate) eta_3: F,
+    pub(crate) beta: F,
+    pub(crate) gamma: F,
+    pub(crate) public_input_delta: F,
+    pub(crate) alphas: [F; NUM_ALPHAS],
+    pub(crate) gate_challenges: Vec<F>,
+}
+
+pub struct GateSeparatorPolynomial<F: PrimeField> {
+    betas: Vec<F>,
+    pub beta_products: Vec<F>,
+    pub partial_evaluation_result: F,
+    current_element_idx: usize,
+    pub periodicity: usize,
+}
+
+impl<F: PrimeField> GateSeparatorPolynomial<F> {
+    pub fn new(betas: Vec<F>, log_num_mononmials: usize) -> Self {
+        let pow_size = 1 << log_num_mononmials;
+        let current_element_idx = 0;
+        let periodicity = 2;
+        let partial_evaluation_result = F::ONE;
+
+        // Barretenberg uses multithreading here and a simpler algorithm with worse complexity
+        let mut beta_products = vec![F::ONE; pow_size];
+        for (i, beta) in betas.iter().take(log_num_mononmials).enumerate() {
+            let index = 1 << i;
+            beta_products[index] = *beta;
+            for j in 1..index {
```

Comment:
This could use some benchmarking if the MT algorithm used by them is really better, or if we can just put a par_iter here and are also good?

Reply:
Yes, but let's move multithreading to a later point